compiler/rustc_const_eval/src/interpret/eval_context.rs

```diff
@@ -1028,11 +1027,12 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 | ty::Closure(..)
                 | ty::CoroutineClosure(..)
                 | ty::Never
-                | ty::Error(_) => true,
+                | ty::Error(_)
+                | ty::Dynamic(_, _, ty::DynStar) => true,
 
-                ty::Str | ty::Slice(_) | ty::Dynamic(..) | ty::Foreign(..) => false,
+                ty::Str | ty::Slice(_) | ty::Dynamic(_, _, ty::Dyn) | ty::Foreign(..) => false,
 
-                ty::Tuple(tys) => tys.last().iter().all(|ty| is_very_trivially_sized(**ty)),
+                ty::Tuple(tys) => tys.last().map_or(true, |&ty| is_very_trivially_sized(ty)),
```

Comment:
I find `map_or` pretty hard to understand, I have to basically inline it to make sense of it. That's why I used `all()`.

`is_none_or` would be the best way to express this, of course (Cc https://github.com/rust-lang/libs-team/issues/212).

Reply:
Interesting. For me it's the exact opposite: I'm confused every time `Option` is used as an iterator and it's not immediately obvious to me that this `.iter().all(...)` checks for at most one element. On the other hand `.map_or(true` already reads as `is_none_or` to me, but that's probably just because I got used to it.

Will revert this change if you prefer.

Reply:
Yeah I think I'd prefer that.